docs/proposals/21-global-failover.md

```diff
@@ -0,0 +1,169 @@
+---
+title: Global Failover
+gep-number: 21
+creation-date: 2022-07-12
+status: implementable
+authors:
+- "@jia-jerry"
+- "@EmoinLanyu"
+reviewers:
+- "@main-reviewer-1"
+- "@main-reviewer-2"
+---
+
+# GEP-21: Fail Over Multi-AZ Services by Isolating Unhealthy AZ
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+    - [New Changes](#new-changes)
+    - [Control Failover / Failback Process](#control-failover--failback-process)
+    - [Failover Workflow](#failover-workflow)
+    - [Failback Workflow](#failback-workflow)
+- [Future Plan](#future-plan)
+
+## Summary
+
+This `GEP` introduces an operating procedure allowing for Gardener operators to isolate an unhealthy AZ of a `Shoot` cluster, and consequently fail over services running inside to healthy AZs. This procedure consists of two steps: removing `MachineDeployment` in the unhealthy AZ, and applying `SubnetIsolation` to the unhealthy AZ. The `SubnetIsolation` is a new extension resource that is reconciled by the respective provider extension controller.
+
+Moreover, it covers a failback procedure for services with multi-AZ setup once the unhealthy AZ is recovered.
+
+## Motivation
+
+Several outages in past years have shown that AZ-specific degradations / failures in hyperscaler services result in significant issues for services. There are multiple patterns of known issues.
+
+1. Some services operate multiple instances of each components redundantly within a single AZ. This will likely provide uptimes approaching 99.5%, but such services are subject to AZ failures and remain offline until the underlying AZ is restored.
+2. Often, multi-AZ HA services are able to stay available during AZ-specific degradations, such as slow disk IO, network lantency, etc. But their api performance may be downgraded tenfold.
+3. In some edge cases, multi-AZ HA services may run into inconsistent state due to unexpected AZ degradations / failures. Manual operations are needed in order to recover the problematic service back to normal.
```

Comment:
Uh, what's that? Poorly written software? I don't think this is a valid use case and even if, I am even more sceptical that with such poorly written software, it would "recover" from "inconsistent state". That sounds like bad software that should not even run if there is no zone outage as even without other things may go wrong and then the software is broken.

Reply:
We have an example. One of our services uses Solace. According to its official document [Recovering from Uncontrolled Fail‑Overs](https://docs.solace.com/Features/DR-Replication/Recov-Uncon-Fail-Over.htm), Solace may run into an inconsistent state during data center outage. We also reproduced such issues in our outage simulation test. As we mentioned in GEP, manual operations are inevitable for recovering services in inconsistent state, however forced failovers of other services can eliminate new inconsistent state during a long period of hyperscaler failures.

Reply:
OK, but this is nuts. Then whoever runs Solace should discontinue using it. "Uncontrolled fail-overs" are exactly the point of fail-overs and if the software is so brittle that it's messed up afterwards, then it generally runs that risk (not tied to zone outage/degredation per se) and can happen any time (point above).

Besides, if the zone has failed already, then the red button is too late anyhow and this thing is already fubar.

Reply:
I agree with you that cloud native services should never run into such state where manual action needs to be involved.  We didn't know such issue in Solace until we tested it. We are not sure whether there are still other services with such issue.

Still, I would say Red button has its value for these services. Always, an outage begins with a smaller scope of infra services in a zone. Then it gradually spreads to a larger scope. This process may take dozens of minutes. Red button should be pressed before things get worse.

Reply:
I (continue to) fail to understand how a well-designed HA solution should be faster/better than any human that presses buttons. Gardener wasn't built this way and I cannot imagine a situation in which I would have wanted a red button for ourselves, so I have severe doubts, this proposal is worth exploring also for everybody on top of Gardener. Fundamentally, Kubernetes gives you all means necessary, so people should use them instead of laying the decision into the hands of people that react slowly and don't have the information available that automats/controllers have. This has the smell of feature creep and nothing else.